sam_view.c

```diff
@@ -404,47 +413,243 @@ int parse_aux_list(auxhash_t *h, char *optarg) {
+    void *mate_names = khash_str2int_init();
+
+    bam1_t *rec = bam_init1();
+    while ( sam_itr_multi_next(conf->in, iter, rec)>=0 )
+    {
+        if ( rec->core.mtid>=0 && bed_overlap(conf->bed, sam_hdr_tid2name(conf->header,rec->core.mtid), rec->core.mpos, rec->core.mpos) ) continue;
+
+        nmates++;
+        if ( !khash_str2int_has_key(mate_names,bam_get_qname(rec)) )
+        {
+            ret = khash_str2int_set(mate_names, strdup(bam_get_qname(rec)), 1);
+            if ( ret<0 ) error("[%s:%d] could not expand memory structures, %d elements. (Todo: external sort)\n",__FILE__,__LINE__,nmates);
+        }
+
+        if ( rec->core.mtid < 0 || (rec->core.flag & BAM_FMUNMAP) ) nunmap = 1;
+        if ( rec->core.mtid >= 0 )
+        {
+            reglist = _reglist_push(reglist, &nreg, sam_hdr_tid2name(conf->header,rec->core.mtid), rec->core.mtid, rec->core.mpos,rec->core.mpos+1);
+            if ( !reglist ) error("[%s:%d] could not expand memory structures, %d elements. (Todo: external sort)\n",__FILE__,__LINE__,nmates);
+        }
+    }
+    _reglist_merge(reglist, nreg);
+    if ( nunmap )
+        reglist = _reglist_push(reglist,&nreg,"*",HTS_IDX_NOCOOR,0,HTS_POS_MAX);
+    hts_itr_multi_destroy(iter);
+    iter = sam_itr_regions(conf->hts_idx, conf->header, reglist, nreg);
+    if ( !iter ) error("[%s:%d] iterator could not be created. Aborting.\n",__FILE__,__LINE__);
+    while ( sam_itr_multi_next(conf->in, iter, rec)>=0 )
+    {
+        int drop = 1;
+        if ( rec->core.tid >=0 && bed_overlap(conf->bed, sam_hdr_tid2name(conf->header,rec->core.tid), rec->core.pos, bam_endpos(rec)) ) drop = 0;
+        if ( drop && khash_str2int_get(mate_names,bam_get_qname(rec),&ret)==0 ) drop = 0;
+        if ( !drop && check_sam_write1(conf->out,conf->header,rec,conf->fn_out,&ret) < 0 ) error("aborting");
+    }
+
+    int i;
+    for (i=0; i<nreg; i++) free((char*)reglist[i].reg);
+    hts_itr_multi_destroy(iter);
+    hts_idx_destroy(conf->hts_idx); // destroy the BAM index
+    conf->hts_idx = NULL;
+    khash_str2int_destroy_free(mate_names);
+    bam_destroy1(rec);
+    return 0;
+}
+
+int multi_region_view(samview_settings_t *conf, hts_itr_multi_t *iter)
+{
+    bam1_t *b = bam_init1();
+    int ret;
+    // fetch alignments
+    while ((ret = sam_itr_multi_next(conf->in, iter, b)) >= 0) {
+        if (!process_aln(conf->header, b, conf)) {
+            if (!conf->is_count) {
+                change_flag(b, conf);
+                if (check_sam_write1(conf->out, conf->header, b, conf->fn_out, &ret) < 0) break;
+            }
+            conf->count++;
+        } else if (conf->unmap) {
+            b->core.flag |= BAM_FUNMAP;
+            if (check_sam_write1(conf->out, conf->header, b, conf->fn_out, &ret) < 0) break;
+        } else {
+            if (conf->un_out) { if (check_sam_write1(conf->un_out, conf->header, b, conf->fn_un_out, &ret) < 0) break; }
+        }
```

Comment:
I think this part is also common to the old-iterator and non-iterator versions, so it would be good to pull it out into its own inline function.